[stable/grafana] Utilize 5.x datasource and dashboard import tooling and general refactor

rtluckie opened the pull request:
- utilize new features of grafana 5.x to configure datasources and dashboards
  - removed all jobs
- update ingress manifest to align with helm create
- remove opinionated affinity config
- consolidate configmaps
- cleanup and refactor values.yaml 
- removed unnecessary nesting from values.yaml
- add initContainer to download dashboards
- update labels and selectors to align with helm best practices
- general cleanup to align with helm best practices/patterns observed in `helm create`
- update values, NOTES, README and _helpers accordingly
- cleanup service manifest to allow more flexibility
- add rtluckie to maintainers
- bump chart version

**NOTE:** This change is not backwards compatible with grafana < 5.0.0

Tested on:
k8s 1.8.10, 1.9.6 and 1.10.0

Limitations:
- Imported dashboards only support a single datasource

Commit:
Refactor grafana

- rename manifest to align with helm create
- remove unnecessary manifests

Commit:
Grafana refactor

- utilize new features of grafana 5.x to configure datasources and dashboards
  - remove all jobs
- update ingress manifest to align with helm create
- consolidate configmaps
- cleanup and refactor values.yaml
- remove unnecessary nesting from values.yaml
- add initContainer to download dashboards
- update labels and selectors to align with helm best practices
- general cleanup to align with helm best practices/patterns observed in `helm create`
- update values, NOTES, README and _helpers accordingly
- cleanup service manifest to allow more flexibility
- add rtluckie to maintainers
- bump chart version

Tested on:
k8s 1.8.10, 1.9.6 and 1.10.0

Limitations:
- Imported dashboards only support a single datasource

rtluckie commented:
/assign @unguiculus

rtluckie commented:
/assign @foxish

DaytonG commented on stable/grafana/templates/deployment.yaml:
```diff
@@ -1,123 +1,134 @@
-          image: busybox
-          command: ['sh', '-c', 'cp /tmp/config-volume-configmap/* /tmp/config-volume 2>/dev/null || true; cp /tmp/dashboard-volume-configmap/* /tmp/dashboard-volume 2>/dev/null || true']
+        - name: download-dashboards
+          image: appropriate/curl
```
Could this also be a setting in values.yaml? Use case there: using a private mirror for deployment

Another little thing: this init container is gonna launch even if we have 0 dashboards to download (perhaps you only provide them via config or retrieve them from grafana via ID). Would be nice if there was some simple way to disable this container or have it only launch when it's necessary.

rtluckie replied:
@DaytonG Sure.  I can make the download-dashboard image a setting in values.yaml

rtluckie replied:
Re: the init container: It was my intention to have it only run when there were dashboards to download, but I haven't been able to figure out how.  I will restructure .Values.dashboards to hopefully make it a bit easier.
Any help would be appreciated!

DaytonG replied:
Hm. The simplest change could be to just move the downloaded dashboards to its own variable and process it independently of the dashboards provided locally. Could potentially increase configuration complexity a little, but it may also simplify some of the template code.

rtluckie replied:
I thought about that, but I'd like to avoid the configuration complexity.  I'll ruminate on it for a while.  Even if the initContainer runs unnecessarily, the image is light (~7mb) and with nothing to download the script is benign.

DaytonG commented on stable/grafana/values.yaml:
```diff
@@ -1,475 +1,148 @@
+#      disableDeletion: false
+#      editable: true
+#      options:
+#        path: /var/lib/grafana/dashboards
```
Perhaps it's just me, but I find that I always need to provision dashboards. It would be ideal if provisioning was enabled by default, so I just needed to fill in `dashboards`

rtluckie replied:
@DaytonG  I think its a general best practice to have minimal sane defaults in values.yaml

DaytonG replied:
Yea, I can see that. My only concern is if someone tried to set `dashboards` but didn't know they _also_ needed to configure providioners it may be difficult to figure out what to do. Perhaps a comment would do for `dashboards` saying "if you need this, you must uncomment the providers configuration"?

rtluckie replied:
Added comment to values.yaml

DaytonG commented on stable/grafana/values.yaml:
```diff
@@ -1,475 +1,148 @@
+    level: info
+  dashboards.json:
+    enabled: true
+    path: /var/lib/grafana/dashboards
```
In G5, `dashboards.json` is deprecated. Since this change is not backwards compatible with G4, could we just nuke the section, for clarity?

rtluckie replied:
@DaytonG Good catch. Fixed

DaytonG commented on stable/grafana/templates/configmap.yaml:
```diff
@@ -1,15 +1,52 @@
+  {{ $key }}: |
+{{ toYaml $value | indent 4 }}
+  {{- end -}}
+{{- end -}}
```
The datasources configuration can contain passwords and other sensitive information. Seems like it should go into a Secret instead of a ConfigMap.

rtluckie replied:
Good call.  grafana.ini should probably also be a secret.  Will fix in follow-up PR.

DaytonG commented on stable/grafana/Chart.yaml:
```diff
@@ -1,5 +1,5 @@
 name: grafana
-version: 0.8.5
+version: 0.9.0
```
Nitpicky: I would argue 1.0.0 is more appropriate here, since this is a backwards incompatible change

rtluckie replied:
Fixed.

DaytonG commented:
(Sorry, I think I may have posted review comments a little wrong. Not used to making them much here on github).

Overall, I think your PR is much better than mine. We should approve yours.

timstoop commented:
@DaytonG in that case, I recommend you close the other one in favour of this one?

Commit:
Changes from comments

- remove deprecated grafana.ini option
- rework .Values.dashboards to make triggering download initContainer easier
- make initContainer image configurable
- add note to values.yaml regarding dashboards' dependency on dashboardProviders
- bump chart version to 1.0.0

timstoop commented:
/ok-to-test

DaytonG commented:
/lgtm

timstoop commented on stable/grafana/templates/NOTES.txt:
```diff
@@ -1,36 +1,35 @@
-     export POD_NAME=$(kubectl get pods --namespace {{ .Release.Namespace }} -l "app={{ template "grafana.fullname" . }},component={{ .Values.server.name }}" -o jsonpath="{.items[0].metadata.name}")
+        You can watch the status of by running 'kubectl get svc --namespace {{ .Release.Namespace }} -w {{ template "grafana.fullname" . }}'
+     export SERVICE_IP=$(kubectl get svc --namespace {{ .Release.Namespace }} {{ template "grafana.fullname" . }} -o jsonpath='{.status.loadBalancer.ingress[0].ip}')
+     echo http://$SERVICE_IP:{{ .Values.httpPort -}}
```
Is the "echo" in this line needed?

rtluckie replied:
@timstoop good catch! fixed.

Commit:
NOTES fixes

- remove unnecessary echo
- fix service.port references

timstoop commented:
/lgtm

unguiculus commented on stable/grafana/Chart.yaml:
```diff
@@ -9,4 +9,6 @@ sources:
 maintainers:
   - name: Ming Hsieh
     email: zanhsieh@gmail.com
+  - name: Ryan Luckie
```
Use Github usernames

rtluckie replied:
fixed.

unguiculus commented on stable/grafana/templates/configmap.yaml:
```diff
@@ -1,15 +1,52 @@
-  {{- if .Values.server.installPlugins }}
-  grafana-install-plugins: {{ .Values.server.installPlugins | quote }}
+{{- if .Values.plugins }}
+  plugins: {{ default "" .Values.plugins | quote }}
```
Don't use the `default` function. Default values should come from `values.yaml`. Apply everywhere.

rtluckie replied:
fixed.

unguiculus commented on stable/grafana/templates/service.yaml:
```diff
@@ -0,0 +1,44 @@
+{{ toYaml . | indent 4 }}
+{{- end }}
+spec:
+{{- if (or (eq .Values.service.type "ClusterIP" "") (empty .Values.service.type)) }}
```
Why would this make sense? `eq .Values.service.type "ClusterIP" ""`

rtluckie replied:
It really doesn't-- it was a relic from refactor.  Fixed.

richerve commented:
This PR is more complete than #3514 although not backwards compatible, I think it should supersede that PR too.

@rtluckie any chance to check @unguiculus review

Commit:
More changes from comments

- update name to gh username in Charts.yaml
- cleanup plugins install
- cleanup dashboards install -- don't rely on `defaults`
- cleanup reusing pv claim -- don't reply on `defaults`
- cleanup secrets -- don't rely on `defaults`
- cleanup service manifest -- don't rely `defaults`
- use storageClassName instead of storageClass (don't rely on `defaults`)
- remove redundant defaults from grafana.ini configmap

rtluckie commented:
@unguiculus @richerve requested changes have been implemented.

unguiculus commented on stable/grafana/templates/deployment.yaml:
```diff
@@ -129,7 +129,11 @@ spec:
       {{- if .Values.persistence.enabled }}
           persistentVolumeClaim:
-            claimName: {{ .Values.persistence.existingClaim | default (include "grafana.fullname" .) }}
+          {{- if .Values.persistence.existingClaim -}}
```
This would be a case when using the `default` function makes sense. Sorry, I probably wasn't clear enough. It shouldn't be used for defaults that should go into `values.yaml`. This is not the case here.

rtluckie replied:
@unguiculus  XD.  I was wondering, but was sticking with your `hard and fast` rule.  Sorry I took comment too literally.  Fixed

Commit:
Fix small regression

- use default when using templated value for pv name when existingClaim
  is not defined

unguiculus commented:
/lgtm

amnk commented:
@rtluckie I always get `Error: YAML parse error on grafana/templates/dashboards-json-configmap.yaml: error converting YAML to JSON: yaml: line 14: did not find expected key` when trying to use this, but I can't figure out what is wrong. My dashboards were working with previous Grafana chart - did we change something with this 1.0 release (except required `json` key)?

**EDIT**: I figured the workaround. Basically, `$RAW_JSON` should be one liner. Otherwise I always get the above output.

rtluckie commented:
@amnk can you post your values.yaml and I will try to help.
Also, please review changes to values.yaml.  Specifically, https://github.com/kubernetes/charts/blob/master/stable/grafana/values.yaml#L111

amnk commented:
@rtluckie Thanks for reply :) yeah, I have that all done. And I've figured out the problem - all dashboards definitions should be one-liners in `$RAW_JSON`, otherwise Helm parser fails. Not sure if it is intended or not, but not vary handy :(

Jokero commented:
@rtluckie Also faced with the same error. 
values.yaml https://gist.github.com/Jokero/abd444598e819c961fc773ca7bccd69d

Jokero commented:
If one-liner solution works, it's really not very convenient

DaytonG commented:
Think I got a fix. In the json-configmap:

```
{{- if .Values.dashboards }}
  {{- range $key, $value := .Values.dashboards }}
    {{- if hasKey $value "json" }}
  {{ $key }}.json: |
{{ $value.json | indent 4 }}  <-- this line is changed
    {{- end }}
  {{- end }}
{{- end -}}
```
Basically, the json needed to be indented properly so the YAML could be parsed.

Will make a PR shortly.

amnk commented:
@DaytonG thanks!

Jokero commented:
Is there any working example of values.yaml? Grafana even does not load dashboards from grafana.net

monotek commented:
https://github.com/kubernetes/charts/blob/master/stable/grafana/values.yaml

Jokero commented:
@monotek It does not work. See my example https://gist.github.com/Jokero/abd444598e819c961fc773ca7bccd69d
I tested without custom json dashboard

monotek commented:
I use it for a week like this.

Edit: Seems you're using Minikube. 
Please make sure you updated your local helm repo via: "helm repo update"

Jokero commented:
Ok. Figured out that I have to use exactly `dashboardProviders.dashboardproviders.yaml` but I changed it to be consistent to `dashboardProviders.dashboardProviders.yaml` (camelCase for both).
